### PR TITLE
fixing ingest empty batch error (SCP-2669)

### PR DESCRIPTION
**Pull request by devonbush**

### CODE REVIEW

File: ingest/expression_files/dense_ingestor.py [hunk 1]
```diff
@@ -28,6 +28,8 @@
 class DenseIngestor(GeneExpression, IngestFiles):
     # ToDo SCP-2635
     ALLOWED_FILE_TYPES = ["text/csv", "text/plain", "text/tab-separated-values"]
+    GENE_BATCH_SIZE = 5
+    DATA_ARRAY_BATCH_SIZE = 1000
```
Comment (jlchang) [inline ingest/expression_files/dense_ingestor.py:32]: Devon, thanks for explaining the intent of `DATA_ARRAY_BATCH_SIZE`. Because this constant isn't intended to limit the number of data arrays that get loaded, it was helpful for me to understand how it was being used.

File: ingest/expression_files/dense_ingestor.py [hunk 2]
```diff
@@ -28,6 +28,8 @@
 class DenseIngestor(GeneExpression, IngestFiles):
     # ToDo SCP-2635
     ALLOWED_FILE_TYPES = ["text/csv", "text/plain", "text/tab-separated-values"]
+    GENE_BATCH_SIZE = 5
```
Comment (bistline) [inline ingest/expression_files/dense_ingestor.py:31]: Out of curiosity, why is `GENE_BATCH_SIZE` so much lower than `DATA_ARRAY_BATCH_SIZE` when the `Gene` documents are so much smaller overall compared to `DataArray` documents?

Comment (knapii-developments) [inline ingest/expression_files/dense_ingestor.py:31]: I don't think we should be batching by gene. The gene documents are very small. Batching by gene will make more calls to the database than necessary. If this is kept the amount of genes should be significantly larger than the amount of data arrays.

Comment (bistline) [reply ingest/expression_files/dense_ingestor.py:31]: +1

File: tests/test_dense.py [hunk 3]
```diff
@@ -177,3 +177,15 @@ def test_transform_fn(self):
...
+        Assures transform function creates gene data model correctly for files
+        with a number of genes that is a multiple of GENE_BATCH_SIZE (SCP-2669)
+        """
+        expression_matrix = DenseIngestor(
+            "../tests/data/dense_matrix_10_genes_15_cells.txt",
+            "5d276a50421aa9117c982845",
+            "5dd5ae25421aa910a723a337",
+        )
+        for actual_gene_models, actual_data_arrays in expression_matrix.transform():
+            self.assertEqual(DenseIngestor.GENE_BATCH_SIZE, len(actual_gene_models))
```
Comment (knapii-developments) [inline tests/test_dense.py:191]: I think this test should check that a total of 'x' cells (an amount equal to the amount of lines in the file -1) were processed as well that the batch size is **less than or equal** to the batch size. As we run into edge cases, these test files are updated to reflect the case. If someone adds additional genes or the amount of genes is not a multiple of 5 this test will fail.

File: ingest/expression_files/dense_ingestor.py [hunk 4]
```diff
@@ -28,6 +28,8 @@
 class DenseIngestor(GeneExpression, IngestFiles):
     # ToDo SCP-2635
     ALLOWED_FILE_TYPES = ["text/csv", "text/plain", "text/tab-separated-values"]
```
Comment (knapii-developments) [inline ingest/expression_files/dense_ingestor.py:30]: Batch sizes values aren't particular to dense matrices but to all expression data. I think these values should move to the GeneExpression class.

File: ingest/expression_files/dense_ingestor.py [hunk 5]
```diff
@@ -228,19 +230,26 @@ def transform(self):
...
+                    f"{str(datetime.datetime.now() - start_time)} elapsed",
+                    extra=self.extra_log_params,
+                )
+                yield gene_models, data_arrays
+                gene_models = []
+                data_arrays = []
+
+        # load any remaining models (this is necessary here since there isn't an easy way to detect the
+        # last line of the file in the iteration above
+        if len(gene_models) > 0:
```
Comment (knapii-developments) [inline ingest/expression_files/dense_ingestor.py:249]: Non blocker but this can be ``` if gene_model```

File: tests/test_dense.py (hunk 3 above)
Comment (devonbush) [reply tests/test_dense.py:191]: The strict equality is actually really important here.  If someone adds more genes to that file (so, e.g. it has 11 genes), it won't actually be testing what we want to test). As the comment says, this is for testing `files with a number of genes that is a multiple of GENE_BATCH_SIZE `  So a failure in this test would correctly tell someone they shouldn't have added genes to that file

File: ingest/expression_files/dense_ingestor.py (hunk 2 above)
Comment (devonbush) [reply ingest/expression_files/dense_ingestor.py:31]: Great question -- that limit on 5 genes was there from before.  I suspect that was a limit we used as a quick-and-dirty batching size when we made the first proof-of-concept of this ingestor, and then just stuck around...

File: tests/test_dense.py (hunk 3 above)
Comment (knapii-developments) [reply tests/test_dense.py:191]: In that event we should still check for the first case scenario. And a comment should be added that the file should not be touched.

Comment (devonbush) [reply tests/test_dense.py:191]: I think what's important for test files is that the name includes the things that are important about the file -- e.g. 10_genes_15_cells  or error_headers.  If you're changing something that requires a name change to the file, that tells you you're changing something a test relies on.  So I'm not worried about this file being touched any more than I'd be worried about, e.g. someone updating the invalid_array_v2.1.2.txt file to not have any arrays.  Someone would have to both ignore the file name AND a failing test in order to make that change.

File: ingest/expression_files/dense_ingestor.py (hunk 5 above)
Comment (devonbush) [reply ingest/expression_files/dense_ingestor.py:249]: I can see this going either way -- `if gene_model` might be more pythonic, but in critical parsing code I'd rather be really explicit about what we are testing.  it's too easy to forget what counts as truthy in Python/Ruby/Javascript (e.g. in javascript `if ([])` evaluates as true) and have bugs creep in as a result

File: ingest/expression_files/dense_ingestor.py (hunk 4 above)
Comment (devonbush) [reply ingest/expression_files/dense_ingestor.py:30]: agree
